<div align="center">

  <a href="https://ultralytics.com/yolov5" target="_blank">
    <img width="1024", src="https://raw.githubusercontent.com/ultralytics/assets/master/yolov5/v70/splash.png"></a>


<br>
  <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a>
  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/segment/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
  <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
<br>

This <a href="https://github.com/ultralytics/yolov5">YOLOv5</a> 🚀 notebook by <a href="https://ultralytics.com">Ultralytics</a> presents simple train, validate and predict examples to help start your AI adventure.<br>See <a href="https://github.com/ultralytics/yolov5/issues/new/choose">GitHub</a> for community support or <a href="https://ultralytics.com/contact">contact us</a> for professional support.

</div>

# Setup

Clone GitHub [repository](https://github.com/ultralytics/yolov5), install [dependencies](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) and check PyTorch and GPU.

In [1]:
!git clone https://github.com/ultralytics/yolov5  # clone
%cd yolov5
%pip install -qr requirements.txt  # install

import torch
import utils
display = utils.notebook_init()  # checks

YOLOv5 🚀 v7.0-33-g1752768 Python-3.8.10 torch-1.13.0+cu117 CUDA:0 (NVIDIA GeForce RTX 3080, 9996MiB)


Setup complete ✅ (32 CPUs, 31.3 GB RAM, 99.3/479.5 GB disk)


# 1. Predict

`segment/predict.py` runs YOLOv5 instance segmentation inference on a variety of sources, downloading models automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases), and saving results to `runs/predict`. Example inference sources are:

```shell
python segment/predict.py --source 0  # webcam
                             img.jpg  # image 
                             vid.mp4  # video
                             screen  # screenshot
                             path/  # directory
                             'path/*.jpg'  # glob
                             'https://youtu.be/Zgi9g1ksQHc'  # YouTube
                             'rtsp://example.com/media.mp4'  # RTSP, RTMP, HTTP stream
```

In [1]:
# After Training with test data, replace --weights with the latest run. 
# Also, replace --source with image to use to detect objects in  
!python yolov5/detect.py --weights yolov5/runs/train/exp11/weights/best.pt --img 640 --conf 0.2 --source yolo_images/test_data/test/3.png
# display.Image(filename='yolov5/runs/predict/exp8/1.png', width=600)

detect: weights=['yolov5/runs/train/exp11/weights/best.pt'], source=yolo_images/test_data/test/3.png, data=yolov5/data/coco128.yaml, imgsz=[640, 640], conf_thres=0.2, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=yolov5/runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
opts:  Namespace(agnostic_nms=False, augment=False, classes=None, conf_thres=0.2, data=PosixPath('yolov5/data/coco128.yaml'), device='', dnn=False, exist_ok=False, half=False, hide_conf=False, hide_labels=False, imgsz=[640, 640], iou_thres=0.45, line_thickness=3, max_det=1000, name='exp', nosave=False, project=PosixPath('yolov5/runs/detect'), save_conf=False, save_crop=False, save_txt=False, source='yolo_images/test_data/test/3.png', update=False, vid_stride=1, view_img=False, visualiz

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<img align="left" src="https://user-images.githubusercontent.com/26833433/199030123-08c72f8d-6871-4116-8ed3-c373642cf28e.jpg" width="600">

# 2. Validate
Validate a model's accuracy on the [COCO](https://cocodataset.org/#home) dataset's `val` or `test` splits. Models are downloaded automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases). To show results by class use the `--verbose` flag.

In [ ]:
# Download COCO val
!bash data/scripts/get_coco.sh --val --segments  # download (780M - 5000 images)

######################################################################## 100.0%
######################################################################## 100.0%


In [ ]:
# Validate YOLOv5s-seg on COCO val
!python segment/val.py --weights yolov5s-seg.pt --data coco.yaml --img 640 --half

segment/val: data=/content/yolov5/data/coco.yaml, weights=['yolov5s-seg.pt'], batch_size=32, imgsz=640, conf_thres=0.001, iou_thres=0.6, max_det=300, task=val, device=, workers=8, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=False, project=runs/val-seg, name=exp, exist_ok=False, half=True, dnn=False
YOLOv5 🚀 v7.0-2-gc9d47ae Python-3.7.15 torch-1.12.1+cu113 CUDA:0 (Tesla T4, 15110MiB)

Fusing layers... 
YOLOv5s-seg summary: 224 layers, 7611485 parameters, 0 gradients, 26.4 GFLOPs
val: Scanning /content/datasets/coco/val2017... 4952 images, 48 backgrounds, 0 corrupt: 100% 5000/5000 [00:03<00:00, 1361.31it/s]
val: New cache created: /content/datasets/coco/val2017.cache
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% 157/157 [01:54<00:00,  1.37it/s]
                   all       5000      36335      0.673      0.517      0.566      0.373

# 3. Train

<p align=""><a href="https://roboflow.com/?ref=ultralytics"><img width="1000" src="https://github.com/ultralytics/assets/raw/master/im/integrations-loop.png"/></a></p>
Close the active learning loop by sampling images from your inference conditions with the `roboflow` pip package
<br><br>

Train a YOLOv5s-seg model on the [COCO128](https://www.kaggle.com/ultralytics/coco128) dataset with `--data coco128-seg.yaml`, starting from pretrained `--weights yolov5s-seg.pt`, or from randomly initialized `--weights '' --cfg yolov5s-seg.yaml`.

- **Pretrained [Models](https://github.com/ultralytics/yolov5/tree/master/models)** are downloaded
automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases)
- **[Datasets](https://github.com/ultralytics/yolov5/tree/master/data)** available for autodownload include: [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [Argoverse](https://github.com/ultralytics/yolov5/blob/master/data/Argoverse.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml), [xView](https://github.com/ultralytics/yolov5/blob/master/data/xView.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [SKU-110K](https://github.com/ultralytics/yolov5/blob/master/data/SKU-110K.yaml).
- **Training Results** are saved to `runs/train-seg/` with incrementing run directories, i.e. `runs/train-seg/exp2`, `runs/train-seg/exp3` etc.
<br><br>

A **Mosaic Dataloader** is used for training which combines 4 images into 1 mosaic.

## Train on Custom Data with Roboflow 🌟 NEW

[Roboflow](https://roboflow.com/?ref=ultralytics) enables you to easily **organize, label, and prepare** a high quality dataset with your own custom data. Roboflow also makes it easy to establish an active learning pipeline, collaborate with your team on dataset improvement, and integrate directly into your model building workflow with the `roboflow` pip package.

- Custom Training Example: [https://blog.roboflow.com/train-yolov5-instance-segmentation-custom-dataset/](https://blog.roboflow.com/train-yolov5-instance-segmentation-custom-dataset/?ref=ultralytics)
- Custom Training Notebook: [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1JTz7kpmHsg-5qwVz2d2IH3AaenI1tv0N?usp=sharing)
<br>

<p align=""><a href="https://roboflow.com/?ref=ultralytics"><img width="480" src="https://robflow-public-assets.s3.amazonaws.com/how-to-train-yolov5-segmentation-annotation.gif"/></a></p>Label images lightning fast (including with model-assisted labeling)

In [ ]:
#@title Select YOLOv5 🚀 logger {run: 'auto'}
logger = 'TensorBoard' #@param ['TensorBoard', 'Comet', 'ClearML']

if logger == 'TensorBoard':
  %load_ext tensorboard
  %tensorboard --logdir runs/train-seg
elif logger == 'Comet':
  %pip install -q comet_ml
  import comet_ml; comet_ml.init()
elif logger == 'ClearML':
  import clearml; clearml.browser_login()

In [2]:
# Train YOLOv5s on COCO128 for 3 epochs
!python yolov5/train.py --img 640 --batch 16 --epochs 1000 --data coco128.yaml --weights yolov5s.pt --cache ram

train: weights=yolov5s.pt, cfg=, data=coco128.yaml, hyp=yolov5/data/hyps/hyp.scratch-low.yaml, epochs=1000, batch_size=16, imgsz=640, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=False, evolve=None, bucket=, cache=ram, image_weights=False, device=, multi_scale=False, single_cls=False, optimizer=SGD, sync_bn=False, workers=8, project=yolov5/runs/train, name=exp, exist_ok=False, quad=False, cos_lr=False, label_smoothing=0.0, patience=100, freeze=[0], save_period=-1, seed=0, local_rank=-1, entity=None, upload_dataset=False, bbox_interval=-1, artifact_alias=latest
remote: Enumerating objects: 227, done.
remote: Counting objects: 100% (174/174), done.
remote: Compressing objects: 100% (46/46), done.
remote: Total 227 (delta 145), reused 154 (delta 128), pack-reused 53
Receiving objects: 100% (227/227), 97.52 KiB | 32.51 MiB/s, done.
Resolving deltas: 100% (164/164), completed with 25 local objects.
From https://github.com/ultralytics/yolov5
   1752768..55


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
      6/999       4.2G    0.07704    0.02154    0.02931         33        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190       0.76     0.0323     0.0255    0.00797

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
      7/999       4.2G    0.07676    0.02248    0.02973         40        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.767     0.0665     0.0284     0.0113

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
      8/999       4.2G    0.07313    0.02115     0.0276         44        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.799      0.115     0.0978     0.0355

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     31/999       4.2G    0.04411    0.01537    0.01682         36        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.513      0.637      0.474      0.289

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     32/999       4.2G    0.04256    0.01543    0.01658         37        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.424      0.658      0.501      0.269

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     33/999       4.2G    0.04452    0.01598    0.01838         47        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190       0.54      0.638      0.494      0.282

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     56/999       4.2G    0.03551    0.01069    0.01221         36        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.777      0.766       0.85      0.577

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     57/999       4.2G    0.03262    0.01139    0.01235         37        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.726      0.796      0.855      0.543

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     58/999       4.2G    0.03913    0.01086    0.01096         37        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.658      0.761      0.787      0.521

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     81/999       4.2G    0.03244    0.01011   0.008304         50        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.964      0.921      0.977      0.415

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     82/999       4.2G    0.03117   0.008492   0.006564         35        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.866      0.929      0.907      0.573

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     83/999       4.2G    0.03403   0.009059   0.006389         44        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.928      0.921      0.968      0.623

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    106/999       4.2G    0.03116    0.00985    0.00483         54        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.954      0.906      0.977      0.734

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    107/999       4.2G    0.02882   0.008417   0.004121         52        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.955      0.899      0.959      0.724

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    108/999       4.2G    0.02966   0.009834   0.005762         32        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.978      0.957      0.989      0.701

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    131/999       4.2G     0.0268   0.007245   0.003977         52        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.983      0.972      0.991      0.774

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    132/999       4.2G    0.02778   0.007283   0.003385         43        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.985      0.961      0.992      0.752

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    133/999       4.2G    0.02724   0.007722   0.003025         46        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.981      0.961      0.992      0.726

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    156/999       4.2G    0.02641   0.007067   0.003031         65        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.961       0.98      0.992       0.81

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    157/999       4.2G    0.02736   0.008414   0.003879         50        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.974      0.913      0.983      0.774

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    158/999       4.2G    0.02548   0.007276   0.004136         39        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.966       0.96       0.99      0.726

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    181/999       4.2G    0.02445   0.007322   0.003165         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.981      0.978      0.993      0.741

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    182/999       4.2G    0.02238   0.006781   0.002848         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.983      0.977      0.993      0.806

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    183/999       4.2G    0.02472   0.006759   0.003077         40        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.982      0.974      0.993      0.762

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    206/999       4.2G    0.02201   0.007104   0.003849         34        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190       0.99      0.983      0.994       0.73

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    207/999       4.2G    0.02388   0.007586   0.003029         48        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.981      0.996      0.994      0.785

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    208/999       4.2G    0.02269    0.00608   0.002272         39        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.983      0.972      0.994      0.807

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    231/999       4.2G    0.02066   0.007062   0.002644         46        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190       0.97      0.986      0.994      0.832

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    232/999       4.2G    0.02177   0.005919   0.003538         49        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.977      0.969      0.994      0.779

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    233/999       4.2G    0.02324   0.006401   0.002688         36        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.962      0.978      0.993       0.83

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    256/999       4.2G    0.02257   0.006901   0.003171         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.993      0.989      0.994      0.795

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    257/999       4.2G    0.02207   0.005199   0.002168         55        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.989      0.991      0.994      0.805

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    258/999       4.2G    0.02255   0.006615   0.003621         54        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.991       0.99      0.994      0.801

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    281/999       4.2G    0.02202   0.005668   0.002392         33        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190       0.99      0.994      0.994      0.782

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    282/999       4.2G    0.01888   0.005872   0.003231         56        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.971      0.991      0.994       0.83

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    283/999       4.2G    0.02054   0.005447   0.002491         33        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.975      0.996      0.994      0.855

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    306/999       4.2G    0.02073   0.005846   0.002159         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.991      0.995      0.994      0.839

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    307/999       4.2G    0.01939   0.005834   0.002627         39        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.988      0.996      0.994      0.865

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    308/999       4.2G    0.02028   0.005676   0.002541         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.992      0.995      0.994       0.86

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    331/999       4.2G    0.01965   0.006119   0.001842         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.992      0.996      0.994      0.832

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    332/999       4.2G    0.01995   0.005291   0.002043         43        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.992      0.995      0.994      0.864

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    333/999       4.2G    0.02118   0.005562   0.001743         47        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.986      0.994       0.86

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    356/999       4.2G    0.01704   0.005225   0.001703         33        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.993      0.975      0.993      0.856

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    357/999       4.2G    0.01975   0.005836   0.002075         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.988       0.99      0.993      0.871

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    358/999       4.2G    0.01939   0.005298   0.001702         39        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.992      0.992      0.993      0.867

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    381/999       4.2G    0.01786   0.005713   0.001891         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190       0.99      0.994      0.994      0.854

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    382/999       4.2G    0.01725   0.005569   0.002062         44        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.987      0.996      0.994      0.851

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    383/999       4.2G     0.0172   0.005607   0.002641         57        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190       0.99      0.987      0.995      0.882

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    406/999       4.2G     0.0187   0.005413   0.001897         56        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.992      0.982      0.993      0.851

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    407/999       4.2G    0.01707   0.005348   0.001856         34        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.987      0.993      0.993       0.88

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    408/999       4.2G    0.01683   0.005381   0.001577         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.984      0.996      0.993       0.89

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    431/999       4.2G    0.01635   0.005092   0.001977         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.981      0.984      0.992      0.876

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    432/999       4.2G    0.01792   0.005364   0.001629         52        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.981      0.983       0.99      0.878

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    433/999       4.2G    0.01645   0.005258   0.001352         37        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.982      0.983      0.991      0.894

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    456/999       4.2G    0.01598   0.005653   0.001091         47        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.891

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    457/999       4.2G    0.01671     0.0046   0.001393         38        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.993      0.996      0.993      0.896

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    458/999       4.2G    0.01542   0.004968   0.001202         35        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.989      0.996      0.993      0.887

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    481/999       4.2G    0.01561    0.00548   0.001824         44        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.994      0.994      0.888

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    482/999       4.2G    0.01572   0.005206   0.001307         46        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.994      0.993      0.906

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    483/999       4.2G    0.01514    0.00419   0.001441         36        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.994      0.993      0.863

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    506/999       4.2G    0.01602   0.004962   0.001778         44        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.994      0.994        0.9

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    507/999       4.2G     0.0156   0.005244   0.001435         50        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.992      0.994      0.909

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    508/999       4.2G      0.014   0.004854   0.001889         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995       0.99      0.994      0.905

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    531/999       4.2G    0.01512   0.004457   0.001734         50        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.995      0.993      0.917

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    532/999       4.2G    0.01369   0.004119   0.001366         36        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.997      0.994      0.993      0.914

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    533/999       4.2G    0.01536   0.004997   0.001578         40        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.996      0.993      0.898

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    556/999       4.2G     0.0141   0.004294   0.001005         49        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.994      0.994      0.925

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    557/999       4.2G    0.01387    0.00389     0.0012         49        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.995      0.994      0.928

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    558/999       4.2G    0.01495    0.00526    0.00141         36        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.995      0.994       0.93

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    581/999       4.2G    0.01386   0.004473    0.00116         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.993      0.994      0.919

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    582/999       4.2G    0.01285    0.00411   0.001383         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.994      0.994      0.925

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    583/999       4.2G    0.01295   0.004325    0.00106         36        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.994      0.926

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    606/999       4.2G    0.01484   0.004572   0.001363         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.994      0.993      0.933

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    607/999       4.2G    0.01232   0.004169   0.001638         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.994      0.993      0.924

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    608/999       4.2G    0.01499   0.005004   0.001265         34        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.932

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    631/999       4.2G    0.01334   0.004258   0.001126         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.924

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    632/999       4.2G    0.01436   0.004322   0.001544         45        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.941

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    633/999       4.2G     0.0128   0.004456   0.001476         46        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.996      0.993      0.934

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    656/999       4.2G    0.01304   0.003841   0.001347         35        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.993      0.993      0.936

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    657/999       4.2G    0.01286   0.004322   0.001632         48        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.994      0.993      0.942

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    658/999       4.2G    0.01161   0.004281   0.001169         55        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.994      0.993      0.934

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    681/999       4.2G     0.0121   0.003816   0.001219         34        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.995      0.993      0.946

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    682/999       4.2G    0.01324   0.004278    0.00167         48        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.996      0.993      0.952

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    683/999       4.2G    0.01163     0.0047    0.00114         65        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.993      0.993      0.947

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    706/999       4.2G    0.01111   0.003813  0.0009904         39        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190       0.99      0.995      0.993      0.944

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    707/999       4.2G    0.01156   0.004221   0.001323         44        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190       0.99      0.996      0.993       0.95

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    708/999       4.2G     0.0123   0.003928   0.001377         53        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.989      0.996      0.993      0.957

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    731/999       4.2G    0.01121   0.003865  0.0009057         51        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.991      0.994      0.993      0.944

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    732/999       4.2G    0.01168   0.003935   0.001735         47        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.991      0.994      0.993       0.94

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    733/999       4.2G    0.01165   0.004047   0.001219         32        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.991      0.994      0.993       0.95

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    756/999       4.2G    0.01067   0.003711  0.0008814         45        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.955

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    757/999       4.2G    0.01026   0.004162  0.0008343         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.996      0.993      0.962

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    758/999       4.2G    0.01111   0.003798   0.001189         46        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.996      0.993      0.961

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    781/999       4.2G    0.01052   0.003456  0.0008015         47        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.963

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    782/999       4.2G    0.01119   0.003943   0.001034         52        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993       0.96

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    783/999       4.2G    0.01093   0.003581   0.001017         34        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.996      0.993      0.953

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    806/999       4.2G    0.01076    0.00404   0.001025         45        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.996      0.993      0.957

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    807/999       4.2G    0.01034   0.003525      0.001         43        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.996      0.993      0.959

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    808/999       4.2G    0.01133   0.003707   0.001134         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.996      0.993      0.958

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    831/999       4.2G   0.009486   0.003379  0.0007354         37        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.995      0.993      0.965

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    832/999       4.2G    0.01037   0.003498   0.001064         39        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.995      0.993      0.958

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    833/999       4.2G   0.009391   0.003613   0.001403         29        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.995      0.993      0.965

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    856/999       4.2G   0.009586   0.003163  0.0008488         49        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.996      0.993      0.965

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    857/999       4.2G    0.00939     0.0034  0.0007898         47        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.963

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    858/999       4.2G    0.01039   0.003828   0.001085         44        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.965

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    881/999       4.2G   0.009862   0.003828  0.0009512         39        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.994      0.996      0.993      0.968

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    882/999       4.2G   0.008849   0.003178  0.0006197         44        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993       0.97

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    883/999       4.2G    0.00932   0.003697   0.001112         50        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993       0.97

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    906/999       4.2G   0.009538    0.00345  0.0009721         46        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993       0.97

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    907/999       4.2G    0.00948   0.003542  0.0008348         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.996      0.993      0.972

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    908/999       4.2G   0.008642   0.003435  0.0006106         51        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.996      0.993      0.968

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    931/999       4.2G   0.009196   0.003405  0.0006808         33        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.996      0.993      0.968

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    932/999       4.2G   0.008776   0.003622  0.0009416         55        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.996      0.993      0.972

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    933/999       4.2G   0.008758   0.003112  0.0006718         38        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.996      0.996      0.993      0.968

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    956/999       4.2G   0.008675   0.003017  0.0008342         44        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.973

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    957/999       4.2G   0.008329   0.003367  0.0006759         46        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.972

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    958/999       4.2G   0.008747   0.003633  0.0005968         42        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.971

      Epo


      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    981/999       4.2G   0.008643   0.003065  0.0007482         41        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.974

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    982/999       4.2G   0.008617   0.003159  0.0009028         56        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.974

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    983/999       4.2G   0.007673   0.002952  0.0004154         43        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all        126        190      0.995      0.996      0.993      0.973

      Epo

# 4. Visualize

## Comet Logging and Visualization 🌟 NEW
[Comet](https://bit.ly/yolov5-readme-comet) is now fully integrated with YOLOv5. Track and visualize model metrics in real time, save your hyperparameters, datasets, and model checkpoints, and visualize your model predictions with [Comet Custom Panels](https://bit.ly/yolov5-colab-comet-panels)! Comet makes sure you never lose track of your work and makes it easy to share results and collaborate across teams of all sizes! 

Getting started is easy:
```shell
pip install comet_ml  # 1. install
export COMET_API_KEY=<Your API Key>  # 2. paste API key
python train.py --img 640 --epochs 3 --data coco128.yaml --weights yolov5s.pt  # 3. train
```

To learn more about all of the supported Comet features for this integration, check out the [Comet Tutorial](https://github.com/ultralytics/yolov5/tree/master/utils/loggers/comet). If you'd like to learn more about Comet, head over to our [documentation](https://bit.ly/yolov5-colab-comet-docs). Get started by trying out the Comet Colab Notebook:
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1RG0WOQyxlDlo5Km8GogJpIEJlg_5lyYO?usp=sharing)

<img width="1920" alt="yolo-ui" src="https://user-images.githubusercontent.com/26833433/202851203-164e94e1-2238-46dd-91f8-de020e9d6b41.png">

## ClearML Logging and Automation 🌟 NEW

[ClearML](https://cutt.ly/yolov5-notebook-clearml) is completely integrated into YOLOv5 to track your experimentation, manage dataset versions and even remotely execute training runs. To enable ClearML (check cells above):

- `pip install clearml`
- run `clearml-init` to connect to a ClearML server (**deploy your own [open-source server](https://github.com/allegroai/clearml-server)**, or use our [free hosted server](https://cutt.ly/yolov5-notebook-clearml))

You'll get all the great expected features from an experiment manager: live updates, model upload, experiment comparison etc. but ClearML also tracks uncommitted changes and installed packages for example. Thanks to that ClearML Tasks (which is what we call experiments) are also reproducible on different machines! With only 1 extra line, we can schedule a YOLOv5 training task on a queue to be executed by any number of ClearML Agents (workers).

You can use ClearML Data to version your dataset and then pass it to YOLOv5 simply using its unique ID. This will help you keep track of your data without adding extra hassle. Explore the [ClearML Tutorial](https://github.com/ultralytics/yolov5/tree/master/utils/loggers/clearml) for details!

<a href="https://cutt.ly/yolov5-notebook-clearml">
<img alt="ClearML Experiment Management UI" src="https://github.com/thepycoder/clearml_screenshots/raw/main/scalars.jpg" width="1280"/></a>

## Local Logging

Training results are automatically logged with [Tensorboard](https://www.tensorflow.org/tensorboard) and [CSV](https://github.com/ultralytics/yolov5/pull/4148) loggers to `runs/train`, with a new experiment directory created for each new training as `runs/train/exp2`, `runs/train/exp3`, etc.

This directory contains train and val statistics, mosaics, labels, predictions and augmentated mosaics, as well as metrics and charts including precision-recall (PR) curves and confusion matrices. 

<img alt="Local logging results" src="https://user-images.githubusercontent.com/26833433/183222430-e1abd1b7-782c-4cde-b04d-ad52926bf818.jpg" width="1280"/>


# Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
- **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


# Status

![YOLOv5 CI](https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([val.py](https://github.com/ultralytics/yolov5/blob/master/val.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/export.py)) on macOS, Windows, and Ubuntu every 24 hours and on every commit.


# Appendix

Additional content below.

In [ ]:
# YOLOv5 PyTorch HUB Inference (DetectionModels only)
import torch

model = torch.hub.load('ultralytics/yolov5', 'yolov5s-seg')  # yolov5n - yolov5x6 or custom
im = 'https://ultralytics.com/images/zidane.jpg'  # file, Path, PIL.Image, OpenCV, nparray, list
results = model(im)  # inference
results.print()  # or .show(), .save(), .crop(), .pandas(), etc.